### ISSUE TITLE
Use mainloop, callbacks (or observer) instead of std::async

### ISSUE THREAD

**Issue by barbieri** [microsoft/cppgraphqlgen#129]
Currently the project uses [`std::async()`](https://en.cppreference.com/w/cpp/thread/async) to execute the resolvers, it may take different execution modes using [`std::launch::deferred`](https://en.cppreference.com/w/cpp/thread/launch):
 - `std::launch::deferred` (default) postpones the execution until `future.get()` is called, blocking the thread execution
 - `std::launch::async` creates a new thread (on Linux it doesn't use thread pools), allows resolvers to work in parallel.

None of them works good in practice, since either you serialize all the work, which is horrible to reach REST, DB or other systems with higher latency; or you create lots of threads in the system (which implies in reaching the kernel, all the thread bookkeeping, etc).

Most production systems will use some kind of "Main Loop" or "Event Loop", that in UNIX will use `select()`, `poll()`, `epoll` (Linux), `kqueue` (MacOS) or variants, which allows the same thread to wait on multiple resources in the same thread.

Then I'd like to see if there is any chance this project will support some kind of callbacks (or observer pattern) that could allow a mainloop/eventloop to be used, or use some primitives such as `setImmediate()` where we can implement a more 
https://promisesaplus.com/ promise, instead of the C++ version, namely allowing `Promise.all()` to be used to get all the fields of an object.

**Comment by wravery** [comment]
There is a way to use custom threading, which is to return your own `std::future` (i.e. waiting for a `std::promise` that gets fulfilled on your own worker) from your accessor methods. The `service::FieldResult` struct is an adapter to convert either regular values or values wrapped in `std::future` into an object that always looks like `std::future`. Earlier versions always returned `std::future`, but that resulted in tons of boilerplate code in all of the accessors which just created and fulfilled a local `std::promise`.

So in this case you'd use `std::launch::deferred` on the top level `resolve`, and then the accessors for your expensive calls would perform their own scheduling and fulfill a `std::promise` (or any other source of `std::future`) asynchronously and in parallel. You can determine on a per-field basis how to handle the scheduling. Using `std::launch::async` is more of a shotgun approach for platforms that use a threadpool by default, or smaller schemas/queries where you can afford to let it fan out on its own (and that's also why it's not the default, you have to opt-in to it).

Longer term I'm thinking about how to leverage co-routines in C++20. I think there's potential for plugging in alternate scheduling like a main/event loop with that model and apply it more consistently while resolving a query. But C++20 is not available widely enough yet to do that, and the initial version of co-routines in C++20 doesn't have enough built-in support in the standard library to make this straightforward.

**Comment by barbieri** [comment]
Hum, interesting. I'll take a look on that approach to launch threads in some specific places. Currently the scheduling and locks are hurting us a lot :-/

**Comment by barbieri** [comment]
closing this one as we'll move to `std::launch::deferred` with thread on our side only.

**Comment by fantasy-peak** [comment]
> There is a way to use custom threading, which is to return your own `std::future` (i.e. waiting for a `std::promise` that gets fulfilled on your own worker) from your accessor methods. The `service::FieldResult` struct is an adapter to convert either regular values or values wrapped in `std::future` into an object that always looks like `std::future`. Earlier versions always returned `std::future`, but that resulted in tons of boilerplate code in all of the accessors which just created and fulfilled a local `std::promise`.
> 
> So in this case you'd use `std::launch::deferred` on the top level `resolve`, and then the accessors for your expensive calls would perform their own scheduling and fulfill a `std::promise` (or any other source of `std::future`) asynchronously and in parallel. You can determine on a per-field basis how to handle the scheduling. Using `std::launch::async` is more of a shotgun approach for platforms that use a threadpool by default, or smaller schemas/queries where you can afford to let it fan out on its own (and that's also why it's not the default, you have to opt-in to it).
> 
> Longer term I'm thinking about how to leverage co-routines in C++20. I think there's potential for plugging in alternate scheduling like a main/event loop with that model and apply it more consistently while resolving a query. But C++20 is not available widely enough yet to do that, and the initial version of co-routines in C++20 doesn't have enough built-in support in the standard library to make this straightforward.
could you add some examples or pseudo code? This is difficult for a novice

**Comment by wravery** [comment]
> could you add some examples or pseudo code? This is difficult for a novice

As of `v4.0`, it does use C++20 coroutines instead of `std::async`, so the example is going to be different from what it looked like when this issue was originally opened. There's a doc explaining some of it here, but it assumes some familiarity with coroutines: [doc/awaitable.md](https://github.com/microsoft/cppgraphqlgen/blob/main/doc/awaitable.md).

There are also some unit tests which exercise each of the included `await_async` launch options in [test/CoroutineTests.cpp](https://github.com/microsoft/cppgraphqlgen/blob/main/test/CoroutineTests.cpp).

Also in the test code, there's an example of implementing a field accessor as a coroutine and using a custom awaitable (by overloading the `co_await` operator for `std::chrono::duration`) to just put the current thread to sleep for 1 second in [Query::getNode](https://github.com/microsoft/cppgraphqlgen/blob/ab07b3508095cab6706f9d7ae579a5e4a2eb2355/samples/today/TodayMock.cpp#L274).

Ultimately, if the platform you're targeting already supports C++20 coroutines (e.g. `cppwinrt` on Windows), then it should be very straightforward to take advantage of that in your field accessors to schedule work on a background thread or a worker queue. If you only have access to the C++20 standard library and pre-C++20 system APIs, then you'll probably need to either settle for one of the included `await_async` launch options, or build your own custom awaitables using the system APIs at your disposal. The [await_worker_queue](https://github.com/microsoft/cppgraphqlgen/blob/ab07b3508095cab6706f9d7ae579a5e4a2eb2355/include/graphqlservice/GraphQLService.h#L165) implementation might be a good starting point for implementing your own, it automatically dispatches coroutines to a single work queue on a dedicated worker thread. If you take the original example of waiting for a system API like `select()`, `poll()`, etc., then you would replace the `std::thread` and the `std::condition_variable` which signals it to wake up with another API that signals a pending coroutine to resume on the current thread.
